### PR TITLE
[AIR/Data] Make `BatchPredictor` lazy

### PR DESCRIPTION
**Pull request by amogkam**
Signed-off-by: amogkam <amogkamsetty@yahoo.com>

Makes `BatchPredictor` lazy by default to support streaming executor.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ericl — APPROVED** [review]

**Review by jianoaix — APPROVED** [review]
Is there any documentation to update?

**Review by c21 — APPROVED** [review]
Can we check all batch prediction notebooks and examples?

File: python/ray/train/batch_predictor.py [hunk 1]
```diff
@@ -101,6 +101,9 @@ def predict(
     ) -> Union[ray.data.Dataset, ray.data.DatasetPipeline]:
         """Run batch scoring on a Dataset.
 
+        .. note::
+            In Ray 2.4, `BatchPredictor` is lazy by default. You must use on of the Datasets consumption APIs, such as iterating through the output, to trigger execution of prediction.
```
Comment (angelinalg) [inline python/ray/train/batch_predictor.py:105]:
```suggestion
            In Ray 2.4, `BatchPredictor` is lazy by default. Use one of the Datasets consumption APIs, such as iterating through the output, to trigger the execution of prediction.
```

**Comment by ericl** [comment]
Are tests ok?

**Comment by amogkam** [comment]
Failing tests are unrelated, going to merge.

**Pull request merged**